edn/src/edn.rustpeg

```diff
@@ -112,48 +156,69 @@ symbol_name = ( symbol_char_initial+ / "." ) ( symbol_char_subsequent* / "." )
 
 keyword_prefix = ":"
 
-pub symbol -> Value =
-    ns:( sns:$(symbol_namespace) namespace_separator {
-        sns
-    })? n:$(symbol_name) {
-        types::to_symbol(ns, n)
-    }
-
-pub keyword -> Value =
-    keyword_prefix ns:( sns:$(symbol_namespace) namespace_separator {
-        sns
-    })? n:$(symbol_name) {
-        types::to_keyword(ns, n)
-    }
-
-pub list -> Value =
-    "(" __ v:(value)* __ ")" {
-        Value::List(LinkedList::from_iter(v))
-    }
-
-pub vector -> Value =
-    "[" __ v:(value)* __ "]" {
-        Value::Vector(v)
-    }
-
-pub set -> Value =
-    "#{" __ v:(value)* __ "}" {
-        Value::Set(BTreeSet::from_iter(v))
-    }
-
-pair -> (Value, Value) =
+pub symbol -> ValueAndSpan =
+    start:#position
+    ns:( sns:$(symbol_namespace) namespace_separator { sns })?
+    n:$(symbol_name)
+    end:#position {
+        ValueAndSpan {
+            inner: SpannedValue::from_symbol(ns, n),
+            span: Span(start, end)
+        }
+    }
+
+pub keyword -> ValueAndSpan =
+    start:#position
+    keyword_prefix
+    ns:( sns:$(symbol_namespace) namespace_separator { sns })?
+    n:$(symbol_name)
+    end:#position {
+        ValueAndSpan {
+            inner: SpannedValue::from_keyword(ns, n),
+            span: Span(start, end)
+        }
+    }
+
+pub list -> ValueAndSpan =
+    start:#position "(" __ v:(value)* __ ")" end:#position {
+        ValueAndSpan {
+            inner: SpannedValue::List(LinkedList::from_iter(v)),
+            span: Span(start, end)
+        }
+    }
+
+pub vector -> ValueAndSpan =
+    start:#position "[" __ v:(value)* __ "]" end:#position {
+        ValueAndSpan {
+            inner: SpannedValue::Vector(v),
+            span: Span(start, end)
+        }
+    }
+
+pub set -> ValueAndSpan =
+    start:#position "#{" __ v:(value)* __ "}" end:#position {
+        ValueAndSpan {
+            inner: SpannedValue::Set(BTreeSet::from_iter(v)),
+            span: Span(start, end)
+        }
+    }
+
+pair -> (ValueAndSpan, ValueAndSpan) =
     k:(value) v:(value) {
         (k, v)
     }
 
-pub map -> Value =
-    "{" __ v:(pair)* __ "}" {
-        Value::Map(BTreeMap::from_iter(v))
+pub map -> ValueAndSpan =
+    start:#position "{" __ v:(pair)* __ "}" end:#position {
+        ValueAndSpan {
+            inner: SpannedValue::Map(BTreeMap::from_iter(v)),
+            span: Span(start, end)
+        }
     }
 
 // It's important that float comes before integer or the parser assumes that
 // floats are integers and fails to parse
-pub value -> Value =
+pub value -> ValueAndSpan =
```

Comment:
Is it not possible to do all of the `#position` stuff here, in `value`?  That is, capture the posititions before and after the `__` markers below, parse `Value` in the internal bits, and then convert to `ValueAndSpan`?  Maybe not, because of the internal whitespace that gets ignored?

Ah, as I think more about this, I think not: the recursive nature of `ValueAndSpan` denies that possibility, since the children of a container need to have the right type.  So sad.  (This might be possible with a `ValueAndSpan<S>` parameterized by an optional span type.)